core/src/main/java/io/temporal/samples/safemessagepassing/ClusterManagerWorkflow.java

```diff
@@ -0,0 +1,169 @@
+
+    @JsonCreator(mode = JsonCreator.Mode.PROPERTIES)
+    public ClusterManagerResult(
+        @JsonProperty("num_currently_assigned_nodes") int numCurrentlyAssignedNodes,
+        @JsonProperty("num_bad_nodes") int numBadNodes) {
+      this.numCurrentlyAssignedNodes = numCurrentlyAssignedNodes;
+      this.numBadNodes = numBadNodes;
+    }
+
+    @JsonProperty("num_currently_assigned_nodes")
+    public int getNumCurrentlyAssignedNodes() {
+      return numCurrentlyAssignedNodes;
+    }
+
+    @JsonProperty("num_bad_nodes")
+    public int getNumBadNodes() {
+      return numBadNodes;
+    }
+  }
+
+  // Be in the habit of storing message inputs and outputs in serializable structures.
+  // This makes it easier to add more overtime in a backward-compatible way.
+  class ClusterManagerAssignNodesToJobInput {
+    // If larger or smaller than previous amounts, will resize the job.
+    private final int totalNumNodes;
+    private final String jobName;
+
+    @JsonCreator(mode = JsonCreator.Mode.PROPERTIES)
+    public ClusterManagerAssignNodesToJobInput(
+        @JsonProperty("total_num_nodes") int totalNumNodes,
+        @JsonProperty("job_name") String jobName) {
+      this.totalNumNodes = totalNumNodes;
+      this.jobName = jobName;
+    }
+
+    @JsonProperty("total_num_nodes")
+    public int getTotalNumNodes() {
+      return totalNumNodes;
+    }
+
+    @JsonProperty("job_name")
+    public String getJobName() {
+      return jobName;
+    }
+  }
+
+  class ClusterManagerDeleteJobInput {
+    private final String jobName;
+
+    @JsonCreator(mode = JsonCreator.Mode.PROPERTIES)
+    public ClusterManagerDeleteJobInput(@JsonProperty("job_name") String jobName) {
+      this.jobName = jobName;
+    }
+
+    @JsonProperty("job_name")
+    public String getJobName() {
+      return jobName;
+    }
+  }
+
+  class ClusterManagerAssignNodesToJobResult {
+    private final Set<String> nodesAssigned;
+
+    @JsonCreator(mode = JsonCreator.Mode.PROPERTIES)
+    public ClusterManagerAssignNodesToJobResult(
+        @JsonProperty("assigned_nodes") Set<String> assignedNodes) {
+      this.nodesAssigned = assignedNodes;
+    }
+
+    @JsonProperty("assigned_nodes")
+    public Set<String> getNodesAssigned() {
+      return nodesAssigned;
+    }
+  }
+
+  @WorkflowMethod
+  ClusterManagerResult run(ClusterManagerInput input);
+
+  @SignalMethod
+  void startCluster();
+
+  @SignalMethod
+  void stopCluster();
+
+  // This is an update as opposed to a signal because the client may want to wait for nodes to be
+  // allocated
+  // before sending work to those nodes.
+  // Returns the list of node names that were allocated to the job.
+  @UpdateMethod
+  ClusterManagerAssignNodesToJobResult assignNodesToJobs(ClusterManagerAssignNodesToJobInput input);
+
+  // Even though it returns nothing, this is an update because the client may want to track it, for
+  // example
+  // to wait for nodes to be unassigned before reassigning them.
```

Comment:
```suggestion
  // Even though it returns nothing, this is an update because the client may want to track it, for
  // example to wait for nodes to be unassigned before reassigning them.
```